Port Covers to Mui2 Part 2

ghzdude opened the pull request:
## What
Ports the rest of the covers to MUI2

## Implementation Details
ALL covers now use MUI2

## Outcome
Covers use MUI2, and old ui code will be removed

ghzdude committed:
initial work on adv energy detector

ghzdude committed:
move ui method to CoverDetectorBase
remove old ui code

ghzdude committed:
port advanced item
improve advanced energy
create GTTextFieldWidget for postfix

ghzdude committed:
improve post fix rendering
improve cursor positioning slightly

ghzdude committed:
port advanced fluid detector
create common ui method in detector base

ghzdude committed:
forgor filter ui
delete old ui code
simplify setters

ghzdude committed:
move filter into column

ghzdude committed:
actually sync isInverted

ghzdude committed:
fix compile and rebase

ghzdude committed:
use default number and add setter method
move parse

ghzdude committed:
fix textfield widgets being too tall
add postfix for conveyor and pump

ghzdude committed:
cleanup + spotless

ghzdude committed:
add todo

ghzdude committed:
port digital interface

ghzdude committed:
make slot text field editable

ghzdude committed:
try improve ender fluid security

ghzdude committed:
delete unused method

ghzdude committed:
keep values proportional instead of resetting on switching modes

ghzdude committed:
fix rebase

ghzdude committed:
fix GTTextFieldWidget and use again

ALongStringOfNumbers commented on src/main/java/gregtech/common/covers/detector/CoverDetectorFluidAdvanced.java:
```diff
@@ -65,62 +70,60 @@ public void renderCover(@NotNull CCRenderState renderState, @NotNull Matrix4 tra
+                                        })))
+                                .child(new ToggleButton()
+                                        .size(72, 18)
+                                        .right(0)
```
right 0? Does this even do anything?

ghzdude replied:
this puts the button to the right of the row

ALongStringOfNumbers commented on src/main/java/gregtech/common/covers/detector/CoverDetectorFluidAdvanced.java:
```diff
@@ -65,62 +70,60 @@ public void renderCover(@NotNull CCRenderState renderState, @NotNull Matrix4 tra
-    private void setMaxValue(String val) {
-        this.max = CoverDetectorBase.parseCapped(val, min + 1, Integer.MAX_VALUE, DEFAULT_MAX);
+    private void setMaxValue(long val) {
+        this.max = clamp((int) val, min + 1, Integer.MAX_VALUE);
```
Why clamp to int max if you are passing in a long?

ALongStringOfNumbers commented on src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:
```diff
@@ -82,78 +87,73 @@ public void update() {
+
+    private void updateWidget(GTTextFieldWidget w) {
+        w.setMaxLength(getLength());
+        w.setNumbers(0, isUsePercent() ? 100 : Integer.MAX_VALUE);
```
Energy is the one thing that I can see being a long in these covers, so why cap at max int?

ghzdude replied:
our text field widget was missing a method to for setting min and max with long values

ALongStringOfNumbers commented on src/main/java/gregtech/common/mui/widget/GTTextFieldWidget.java:
```diff
@@ -0,0 +1,259 @@
+            super.draw(text, x, y);
+        }
+
+        private static boolean isNumber(String text) {
```
This is unused.

ALongStringOfNumbers commented on src/main/java/gregtech/common/mui/widget/GTTextFieldWidget.java:
```diff
@@ -0,0 +1,259 @@
+import java.util.function.Supplier;
+import java.util.regex.Pattern;
+
+public class GTTextFieldWidget extends BaseTextFieldWidget<GTTextFieldWidget> {
```
A decent amount of this class is unused. Do we need to keep any of it?

ghzdude replied:
the methods are there to match mui2's TextFieldWidget. I could extend TextFieldWidget and remove the redundant methods, but then builder calls become more complicated due to methods returning TextFieldWidget instead of GTTextFieldWidget

ALongStringOfNumbers commented on src/main/java/gregtech/common/mui/widget/GTTextFieldWidget.java:
```diff
@@ -0,0 +1,259 @@
+        }
+        setText(this.stringValue.getStringValue());
+        if (!hasTooltip()) {
+            tooltipBuilder(tooltip -> tooltip.addLine(IKey.str(getText())));
```
Why are we adding a tooltip if there is no tooltip?

ghzdude committed:
clamp values when switching modes

ALongStringOfNumbers commented on src/main/java/gregtech/common/covers/ender/CoverAbstractEnderLink.java:
```diff
@@ -97,6 +98,11 @@ protected final UUID getOwner() {
+        return this.playerUUID == null || this.playerUUID.equals(other) ||
+                FTBTeamHelper.isSameTeam(this.playerUUID, other);
+    }
+
```
Is the KeyPressed Override in this class an option to move to our GTTextField, and then just handle behavior based on if a sync handler is passed in to the method?

ALongStringOfNumbers reviewed (commented):
The tooltip of the continuous/latched mode in the Advanced Fluid Detector needs a new line, it is much too long.

Similar to a previous issue brought up, configuring the filter and then removing it will apply its configuration status to any new filter put in the Advanced Fluid Detector Cover.

Some how I was able to have filter configuration not saved when I exited the GUI of the Advanced Fluid Detection cover. This only happened once and I could not repeat it.

This error is back:
`[20:43:41] [Client thread/INFO] [modularui]: unit WIDTH of widget Column was already used and will be overwritten with unit RIGHT`
So maybe we need more debug names

Same filter issues with item filters in the Advanced Item detector. Putting in a smart filter, removing it and putting in an oredict filter, and then opening the filter UI still opens the smart filter.

Just a little weird math probably, when placing an advanced energy detection cover on a creative emitter, the default value for max is 488%

The Ender Link Cover does not work with the Creative Tank in my testing.

ghzdude committed:
do not attach for creative emitter

ghzdude committed:
fix unit overwrite and redundant method call
missed `coverChildrenHeight()` in CoverDetectorFluidAdvanced

ghzdude committed:
return the creative handler(s) in capability directly
fix oversight from filter being set to error instead of null

ghzdude committed:
delete deprecated ui code/classes

ghzdude committed:
improve latch tooltip

ghzdude commented on src/main/java/gregtech/integration/ftb/utility/FTBTeamHelper.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.UUID;
+
+public class FTBTeamHelper {
```
it isn't really related to chunks, perhaps those methods could be moved here?

ghzdude committed:
fix incorrect postfix

ghzdude committed:
add missing long method to GTTextFieldWidget
add missing tooltip override usage
use proper supplier interfaces
use MathHelper clamp

ghzdude committed:
delete unused method

ghzdude committed:
replace anonymous class with GTTextFieldWidget
add string consumer when enter is pressed

ghzdude committed:
delete deprecated filter ui code

ghzdude committed:
make panel handler based on stack translation key
spotless

ghzdude commented:
> Similar to a previous issue brought up, configuring the filter and then removing it will apply its configuration status to any new filter put in the Advanced Fluid Detector Cover.

> Same filter issues with item filters in the Advanced Item detector. Putting in a smart filter, removing it and putting in an oredict filter, and then opening the filter UI still opens the smart filter.

both of these are caused by the same problem; synced panel handlers only open their panel once. any further attempts to open them will just reopen the cached panel. deleting the cached panel is not an option because `SimpleItemFilter` uses phantom item slots, which throws an exception. I can't use a client only panel handler unless i replace the phantom item slots with something else. I could use a different mapkey

ghzdude committed:
nvm, that doesn't work

ghzdude committed:
brachy is gonna hate this
fixes changing the filter in the filter slot by making a new panel handler and utilizing sync actions and `getOrCreateSyncHandler()`
this is what 330am brain comes up with
even god lives in fear of this abomination

ghzdude committed:
add todo

ghzdude committed:
clean up a bit

ghzdude committed:
use predicate

ghzdude committed:
fix cover filters not saving and syncing allowFlow
fix CoverItemFilter filtermode syncing

ghzdude committed:
todo

ghzdude committed:
spotless?

ghzdude committed:
properly support long counts for items and fluids

Zorbatron commented on src/main/java/gregtech/api/cover/CoverUIFactory.java:
Wasn't the plan with MUI0 to only deprecate it in 2.9, not remove it completely?

ALongStringOfNumbers replied:
This is true. There are addons out there that add covers (Will probably be broken anyway by the update), so it would be good to keep these classes around for a bit and just mark them as deprecated-for-removal

Zorbatron commented on src/main/java/gregtech/api/cover/CoverWithUI.java:
```diff
@@ -30,28 +29,17 @@
+import java.util.function.Supplier;
 
-    @ApiStatus.Experimental
-    default boolean usesMui2() {
```
Ditto here.

Zorbatron commented on src/main/java/gregtech/common/covers/CoverDigitalInterface.java:
```diff
@@ -451,52 +483,90 @@ public boolean modeLeftClick(EntityPlayer entityPlayer, MODE mode, int slot) {
+                                }))
+                        .child(new GTTextFieldWidget()
+                                .setNumbers(0, Integer.MAX_VALUE)
+                                .value(new StringValue.Dynamic(
```
You could put `slotValue` here directly and remove the manual sync manager registration above.

Zorbatron commented on src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:
```diff
@@ -82,78 +87,76 @@ public void update() {
+                                        .size(72, 18)
+                                        .addTooltipLine(IKey.lang("cover.advanced_energy_detector.modes_tooltip"))
+                                        .value(new BooleanSyncValue(this::isUsePercent, this::setUsePercent))
+                                        .overlay(new DynamicDrawable(() -> IKey
```
A `DynamicDrawable` isn't needed here, `LangKey`s can be dynamic (and always use a supplier internally too).

Zorbatron commented on src/main/java/gregtech/common/covers/detector/CoverDetectorFluidAdvanced.java:
```diff
@@ -65,62 +72,68 @@ public void renderCover(@NotNull CCRenderState renderState, @NotNull Matrix4 tra
+                                        .size(72, 18)
+                                        .value(new BooleanSyncValue(this::isInverted, this::setInverted))
+                                        .addTooltipLine(IKey.lang("cover.generic.advanced_detector.invert_tooltip"))
+                                        .overlay(new DynamicDrawable(() -> {
```
Ditto about dynamic lang keys here too.

Zorbatron commented on src/main/java/gregtech/common/covers/detector/CoverDetectorItemAdvanced.java:
```diff
@@ -53,74 +60,62 @@ public void renderCover(@NotNull CCRenderState renderState, @NotNull Matrix4 tra
+                                .child(new ToggleButton()
+                                        .name("inverted button")
+                                        .size(72, 18)
+                                        .overlay(new DynamicDrawable(() -> {
```
Dynamic lang key.

Zorbatron commented on src/main/java/gregtech/common/covers/detector/CoverDetectorFluidAdvanced.java:
Same thing about the filled % bar as the advanced energy detector for the advanced fluid detector. It could also be done to the advanced item detector.

Zorbatron reviewed (commented):
~~I see a few methods in covers to build repetitive widgets, and they have `GuiData` instead of `PosGuiData`.~~

ghzdude committed:
replace dynamic drawables with supplier lang key
use slotValue directly as value

ghzdude committed:
use flow as return type

ghzdude commented:
> I see a few methods in covers to build repetitive widgets, 

what methods are you talking about exactly?
>and they have `GuiData` instead of `PosGuiData`.

I use `GuiData` in case we want to build the widgets in a different context, see #2658

ghzdude committed:
use throughput directly as value

ghzdude committed:
spotless

ALongStringOfNumbers commented on src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:
```diff
@@ -82,78 +86,78 @@ public void update() {
+                                        .overlay(IKey.lang(() -> {
+                                            String lang = "cover.advanced_energy_detector.mode_";
+                                            lang += isUsePercent() ? "percent" : "eu";
+                                            return lang;
```
Would it be a good idea to construct these lang keys in advance? Same with the inverted one.

ALongStringOfNumbers commented on src/main/java/gregtech/api/cover/CoverWithUI.java:
```diff
@@ -104,11 +91,22 @@ default void markAsDirty() {
+                        .overlay(new DynamicDrawable(() -> itemDrawable.setItem(stack.get())))
+                        .size(16)
+                        .marginRight(4))
+                .child(IKey.dynamic(() -> stack.get().getDisplayName())
```
Does the name need to be dynamic? The only case I can think of for this is changing langs, but you have to close the UI to do that so the UI would get rebuilt with the correct lang even if it was not dynamic right?

ghzdude replied:
this was part of my attempts to fix the filter popup panel before settling with the current solution

since the panel is only built once, it needs to update based on what's in the filter slot, therefore it needs a supplier

i could revert this change, though i'd probably add it back later in a different pr

ALongStringOfNumbers commented on src/main/java/gregtech/common/covers/detector/CoverDetectorFluidAdvanced.java:
```diff
@@ -65,62 +71,68 @@ public void renderCover(@NotNull CCRenderState renderState, @NotNull Matrix4 tra
+                                        .size(72, 18)
+                                        .right(0)
+                                        .overlay(IKey.lang(() -> {
+                                            String lang = "cover.generic.advanced_detector.";
```
Could the numbers in these lang keys be colorized so that they stand out more?

ALongStringOfNumbers commented on src/main/java/gregtech/common/covers/detector/CoverDetectorFluidAdvanced.java:
```diff
@@ -162,34 +174,34 @@ public void update() {
-        this.min = tagCompound.getInteger("min");
-        this.max = tagCompound.getInteger("max");
+        this.min = tagCompound.getLong("min");
+        this.max = tagCompound.getLong("max");
```
Do you know if this will mess up existing setups? What I mean is that previously min/max were ints, and I don't know if `readLong` will see the Int values that were previously there, or if it will only look for a new tag. I think it should be ok, because of using the same key, but I am not sure.

ghzdude replied:
in the tag compound, all `getNumber()` methods use type 99 "TAG_ANY_NUMERIC". i don't think it should be an issue, but it's probably better to be safe than sorry to explicitly check for it

ALongStringOfNumbers commented on src/main/java/gregtech/common/covers/filter/BaseFilter.java:
```diff
@@ -34,7 +34,7 @@ public BaseFilterReader getFilterReader() {
 
         @Override
-        public @NotNull ModularPanel createPopupPanel(PanelSyncManager syncManager) {
+        public @NotNull ModularPanel createPopupPanel(PanelSyncManager syncManager, String panelName) {
```
I feel like this panel name could be used here for the error panel.

ALongStringOfNumbers reviewed (commented):
You might also want to make the item/fluid detectors not able to attach to their respective creative sources, to mimic the energy cover.

When clicking on a slot in a fluid filter with a filled bucket, if the slot is already configured to a different fluid, the first click with the bucket will instead remove the configured fluid. You then have to click again to set the fluid in the slot to the one in the bucket. This should be condensed to one operation. If you are clicking on a configured fluid slot with a bucket that contains a different fluid, that fluid slot should just be set to the fluid in the bucket.

Testing this in the Advanced Fluid Detector cover. If you have a filter open, click on an item to open JEI for an item, close JEI, then you can no longer close the filter with the gear button in the main GUI, you have to use the `x` in the filter GUI. When the filter is in this state, I also cannot configure it using fluids from the inventory, only with JEI drag ghosts. However, after closing and reopening the filter, none of these ghost drags made when the filter was in this state are saved. Also you can't click to remove configured fluids.

In the Advanced Energy Detector cover, switching from the default discrete mode to percentage mode and then back sets the min EU value to `-1`. This was on the UHV 8x Battery Buffer. We should probably validate that the EU value in the text field cannot be negative.

Just noticed that the Quantum Network "Not Connected" TOP infor displays when looking at covers on the quantum tank/chest. That can be fixed in a different PR though.

When holding a stack of filters, right clicking a filter slot in a GUI to deposit 1 filter sets the NBT on the entire stack. instead of just the one that went in the slot. Not sure if anything can be done about this one.

ghzdude committed:
add CoverUIFactory back and mark for removal
add notes

ghzdude committed:
make item/fluid detectors unplaceable on creative chest/tank

ghzdude committed:
drain and fill at the same time for our phantom fluid slots
move locking call into sync handlers' handle click method on server

ghzdude committed:
use default values when energy capacity is zero
prevent redundant sets

ghzdude committed:
deserialize legacy int min and max

ghzdude committed:
use passed in name for error panel

ghzdude committed:
simplify lang key overlay

ghzdude committed:
add color to the numbers

ghzdude commented:
> Testing this in the Advanced Fluid Detector cover. If you have a filter open, click on an item to open JEI for an item, close JEI, then you can no longer close the filter with the gear button in the main GUI, you have to use the `x` in the filter GUI. When the filter is in this state, I also cannot configure it using fluids from the inventory, only with JEI drag ghosts. However, after closing and reopening the filter, none of these ghost drags made when the filter was in this state are saved. Also you can't click to remove configured fluids.

Looking into this, this happens because the panel handler on the client thinks the panel isn't opened. However, trying to open the panel again never sets open to true, so it can't close the panel. also in this state, the popup panel is essentially nonfunctional as it cannot receive packet updates on the server (it no longer exists in ModularSyncManager). the x button works because it calls closePanel directly.

in short, mui2 desync issue

> When holding a stack of filters, right clicking a filter slot in a GUI to deposit 1 filter sets the NBT on the entire stack. instead of just the one that went in the slot. Not sure if anything can be done about this one.

can you elaborate on this? i don't know exactly what you're trying to do

ghzdude committed:
don't handle stack tag in constructor
implement tank interface instead of extending
add todo

ghzdude committed:
really

ghzdude committed:
not sponsored by AMD™

Zorbatron commented on src/main/java/gregtech/api/util/virtualregistry/VirtualEnderRegistry.java:
```diff
@@ -91,18 +90,6 @@ private static VirtualRegistryMap getRegistry(UUID owner) {
         return VIRTUAL_REGISTRIES.computeIfAbsent(owner, key -> new VirtualRegistryMap());
     }
 
-    // remove if tank app is removed
```
Was HF going to reimplement the tank app in her pr, or is it not needed since the ender covers have a channel browser built into them now.

Zorbatron commented on src/main/java/gregtech/common/covers/detector/CoverDetectorEnergy.java:
```diff
@@ -27,7 +28,9 @@ public CoverDetectorEnergy(@NotNull CoverDefinition definition, @NotNull Coverab
 
     @Override
     public boolean canAttach(@NotNull CoverableView coverable, @NotNull EnumFacing side) {
-        return coverable.getCapability(GregtechCapabilities.CAPABILITY_ENERGY_CONTAINER, null) != null ||
```
Can the below TODO be done now? ("check this", plz write descriptive todos smh (not that you wrote that one ghz))

ghzdude replied:
iirc, that todo was added by maya, idk what she means by it

Zorbatron commented on src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:
```diff
@@ -82,104 +85,107 @@ public void update() {
+                                                .style(IKey.WHITE))
+                                        .overlay(false, IKey.lang("cover.advanced_energy_detector.mode_eu")
+                                                .style(IKey.WHITE)))))
+                .child(Flow.row()
```
This needs to be nested under the column flow.
<img width="382" height="282" alt="image" src="https://github.com/user-attachments/assets/5a3b3a8b-61de-47cc-9b72-fa3d48227235" />

Zorbatron commented on src/main/java/gregtech/common/covers/detector/CoverDetectorEnergy.java:
This isn't really relevant to this PR, but is it correct that in "normal" mode, 0% = 15 redstone and 100% = 0 redstone? It feels like it's the wrong way around. Same for the item and fluid detector covers.

ghzdude replied:
the original use case that the advanced energy detector was made for was stopping turbines once a battery buffer was full. machine controller by default disable machines when given redstone, so i made the advanced energy detector return zero when past the max value to make it work with machine controllers by default.

advanced item/fluid detectors were then refactored to share the same logic as the advanced energy detector

Zorbatron replied:
Makes sense. It just felt counter intuitive at first.

Zorbatron replied:
Makes sense, it just felt counter intuitive at first.

Zorbatron replied:
Ok github

Zorbatron commented on src/main/java/gregtech/common/covers/filter/BaseFilterContainer.java:
```diff
@@ -226,9 +235,16 @@ public IWidget initUI(GuiData data, PanelSyncManager manager) {
                                         panel.closePanel();
                                     }
+                                    if (client) {
+                                        manager.callSyncedAction("update_filter_panel", packetBuffer -> {});
```
I can't commit it since it would invalidate an approval from me, but once MUI 3.0.6 releases ~~pr 183 merges~~, the empty lambda can be removed.

ghzdude committed:
i accidentally the parenthesis

ghzdude commented on src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:
i did have ideas like that for a while, something like create. it's best in a future pr though
